src/erleans_grain.erl

```diff
@@ -362,115 +373,73 @@ upd_timer(refresh_timer, DeactivateAfter) ->
     {keep_state, Data#data{cb_state=NewCbData}, Actions};
-handle_reply_({stop, NewCbData}, Data, _Actions) ->
-    finalize_and_stop(Data#data{cb_state=NewCbData}, false);
-handle_reply_({save, Updates, NewCbData}, Data=#data{id=Id,
-                                                     cb_module=CbModule,
-                                                     provider=Provider,
-                                                     etag=ETag}, Actions) when is_map(Updates) ->
-    NewETag = update_state(CbModule, Provider, Id, Updates, NewCbData, ETag),
-    {keep_state, Data#data{cb_state=NewCbData, etag=NewETag}, Actions};
-handle_reply_({save, NewCbData}, Data=#data{id=Id,
-                                            cb_module=CbModule,
-                                            provider=Provider,
-                                            etag=ETag}, Actions) ->
-    NewETag = replace_state(CbModule, Provider, Id, NewCbData, ETag),
-    {keep_state, Data#data{cb_state=NewCbData, etag=NewETag}, Actions}.
-
-%% Saving requires an etag so it can verify no other process has updated the row before us.
-%% The actor needs to crash in the case that the etag found in the table has changed.
-update_state(_CbModule, undefined, _Id, _Updates, _CbData, _ETag) ->
-    exit(?NO_PROVIDER_ERROR);
-update_state(CbModule, Provider, Id, Updates, #{persistent := PData,
-                                                ephemeral  := _}, ETag) ->
-    NewETag = etag(PData),
-    update_state_(CbModule, Provider, Id, Updates, ETag, NewETag);
-update_state(CbModule, Provider, Id, Updates, CbData, ETag) ->
-    NewETag = etag(CbData),
-    update_state_(CbModule, Provider, Id, Updates, ETag, NewETag).
-
-update_state_(CbModule, {Provider, ProviderName}, Id, Updates, ETag, NewETag) ->
-    case Provider:update(CbModule, ProviderName, Id, Updates, ETag, NewETag) of
-        ok ->
-            NewETag;
-        {error, Reason} ->
-            exit(Reason)
-    end.
-
-replace_state(_CbModule, undefined, _Id, _CbData, _ETag) ->
+handle_result({ok, NewCbData, CbActions}, Data=#data{ref=GrainRef}, Actions) ->
+    maybe_enqueue_grain(GrainRef),
+    {Actions1, Data1} = handle_actions(CbActions, Actions, NewCbData, Data),
+    {keep_state, Data1#data{cb_state=NewCbData}, Actions1};
+handle_result({deactivate, NewCbData}, Data, _) ->
+    {next_state, deactivating, Data#data{cb_state=NewCbData}, []};
+handle_result({deactivate, NewCbData, CbActions}, Data, _) ->
+    {Actions1, Data1} = handle_actions(CbActions, [], NewCbData, Data),
+    {next_state, deactivating, Data1#data{cb_state=NewCbData}, Actions1}.
+
+%% Drops unrecognized actions and converts cast actions to next_events that
+%% do not reset the activation timer
+handle_actions([], ActionsAcc, _CbData, Data) ->
+    {lists:reverse(ActionsAcc), Data};
+handle_actions([{cast, Msg} | Rest], ActionsAcc, CbData, Data) ->
+    handle_actions(Rest, [{next_event, cast, {leave_timer, Msg}} | ActionsAcc], CbData, Data);
```

Comment:
I don't understand why casts don't reset the activation timer.

Reply:
It isn't the fact that it is a cast that matters, it is because it is a returned "next event", not a request coming in. Same for `info`. And they can't be controlled by the pdict, that is used for when a timer makes a request externally and it is used when sending the request to the server, this is internal.